### ISSUE TITLE
[analyzer] Computation of `UP` with an intersection type seems to omit a case

### ISSUE THREAD

**Issue by eernstg** [dart-lang/sdk#54208]
Consider the following program:

```dart
var b = true;

void f<X extends num?, Y extends X>(X x, Y y) {
  if (x != null) { // `x` is promoted to `X & num`.
    // UP(X & num, Y) == X, because `Y <: X`.
    var v1 = b ? x : y;
    v1.expectStaticType<Exactly<X>>; // Error!
  }
}

void main() {}

// Helpers.

typedef Exactly<X> = X Function(X);

extension<X> on X {
  X expectStaticType<Y extends Exactly<X>>() => this;
}
```

This program causes `UP(X & num, Y)` to be computed in order to find the type of `v1`. This computation should invoke the second item in the definition [here](https://github.com/dart-lang/language/blob/main/resources/type-system/upper-lower-bounds.md#:~:text=(T2)-,UP(X1%20%26%20B1%2C%20T2)%20%3D,-T2%20if%20X1), that is

> otherwise `X1` if `T2 <: X1`

which is true in this case (where `X1` is `X`, `T2` is `Y`, and `Y extends X`). However, the analyzer reports an error at the comment 'Error!', and claims (indirectly) that the type of `v1` is `num?` (that is, the bound of `X`).

The rule which is being used here has been [moved](https://github.com/dart-lang/sdk/issues/54002) from one position in the definition of `UP` to an earlier position, but that should not make a difference for this particular computation, it should give `v1` the type `X` both before and after this change.

**Comment by eernstg** [comment]
@pq and @scheglov, perhaps this was fixed already and should be closed? I don't see the error message at the line with the comment 'Error!' any more (using DartPad based on Dart SDK 3.3.0-238.0.dev).

**Comment by pq** [comment]
@scheglov: did you fix this already? (Maybe along w/ the fix for #54454?)

**Comment by scheglov** [comment]
Yes, I don't see this error too.
Not 100% sure who fixed it, you with your change of UP, or me with other changes related to extension types.
